code-studio/src/js/levels/dialogHelper.js

```diff
@@ -157,58 +159,68 @@ window.dashboard.dialog = (function () {
    * @param {function(boolean)} onComplete Optional callback function to call when
    *     the server call completes, which receives a boolean indicating whether
    *     the browser will redirect to a new location after it is called.
+   * @param {function(function)} beforeHook Optional callback function to call
+   *     before processResults gets underway. This function must call the completion
+   *     function passed as the parameter to complete the processResults call.
    */
   // TODO(dave): move this logic into appOptions.onAttempt for levels of type
```

Comment:
I'm a little worried the lifecycle of `DSLDefined` levels is getting pretty complex.  Rather than rolling a new `beforeHook` now might be the right to address this TODO?

We don't need to completely unify with [`initApp.js`](https://github.com/code-dot-org/code-dot-org/blob/d4b5352b60f0a223ada76b41e983b4e0305e3629/code-studio/src/js/initApp/initApp.js#L54-L82) but it would be good to take an approach that brings these two worlds closer together vs. further apart.